### ISSUE TITLE
Exit with a non-zero status and don't emit on error

### ISSUE THREAD

**Issue by caspervonb** [microsoft/TypeScript#37011]
**TypeScript Version:**  3.7.5

**Search Terms:**

**Code**

```ts
let 1 = 1;
```

**Expected behavior:**
Error is reported and _tsc_ exits with a non zero status and skips emitting the target output file.

**Actual behavior:**

Error is reported but tsc exits with a status code of zero and writes a garbage file to disk. This makes it impossible to figure out if the build succeeded or not from a shell or build tools like _make_.

**Playground Link:** 

https://www.typescriptlang.org/play/index.html?target=1&module=3#code/DYUwLgBAjBC80G4g

**Related Issues:** 

https://github.com/microsoft/TypeScript/issues/4095

**Comment by AnyhowStep** [comment]
--noEmitOnError

https://www.typescriptlang.org/docs/handbook/compiler-options.html

**Comment by RyanCavanaugh** [comment]
`noEmitOnError` will control the emit behavior; please provide a repro for the exit code behavior you're seeing (I see an exit code of 2 here)

**Comment by caspervonb** [comment]
> noEmitOnError will control the emit behavior;

Fair enough, why this isn't the default behavior tho? Fairly odd that a compiler produces garbage on errors by default. Warning; fair enough but on errors?

Am unable to reliably reproduce the exit code bug; it's happening *sometimes* so closing this til I run into it again.

**Comment by AnyhowStep** [comment]
TS does not distinguish between warnings and errors. TS only has errors.

A use case for `noEmitOnError : false` is for gradual typing. Where your code isn't 100% correct TS on day 1, but you need to keep releasing features for your application.

**Comment by caspervonb** [comment]
> TS does not distinguish between warnings and errors.

It makes sense for type checks but there are cases where it will just not emit a node on an error.

For example, on failure to resolve a module, the entire import statement is removed but _tsc_ still outputs the target file.

In other words, it seems like there are two classes of errors, fatal ones where code is omitted from the output and non fatal ones where code is left as is.
